Mathlib/Algebra/Homology/ShortComplex/ExactFunctor.lean

```diff
@@ -46,6 +47,244 @@ noncomputable def preservesFiniteColimitsOfPreservesHomology
+    haveI := preservesEpimorphism_of_preserves_shortExact_right F h
+    refine ⟨?_, inferInstance⟩
+    have := factorThruImage S.g
+    let s : ShortComplex C := .mk (image.ι S.f : image S.f ⟶ S.X₂) S.g $ by
+      simp [← cancel_epi (factorThruImage S.f)]
+
+    have se : s.ShortExact :=
+    { exact := (by
+
+        rw [ShortComplex.exact_iff_kernel_ι_comp_cokernel_π_zero] at e1 ⊢
+        rw [show S.f = _ from image.fac S.f |>.symm] at e1
+        simpa using e1 =≫ (cokernelEpiComp _ _).hom)
+      mono_f := inferInstance
+      epi_g := inferInstance }
+    have := (s.map F).exact_and_epi_g_iff_g_is_cokernel.1 (h _ se) |>.some
+    apply ShortComplex.exact_of_g_is_cokernel
+    simp only [ShortComplex.map_X₂, ShortComplex.map_X₃, ShortComplex.map_g, ShortComplex.map_X₁,
+      ShortComplex.map_f] at this ⊢
+    apply isCokernelEpiComp (i := this) (g := F.map $ factorThruImage S.f)
+    simp [← F.map_comp]
+
+  tfae_have 2 → 3
+  · intro h X Y f
+    refine ⟨⟨fun {c} hc => ?_⟩⟩
+
+    have epi0 : Epi (c.ι.app .one) := epi_of_isColimit_cofork hc
+    let s : ShortComplex C := .mk f (c.ι.app .one) $ by simp
+    have exact0 : s.Exact := by
+      refine ShortComplex.exact_of_g_is_cokernel _ $
+        Limits.IsColimit.equivOfNatIsoOfIso (Iso.refl _) _ _ ⟨⟨?_, ?_⟩, ⟨?_, ?_⟩, ?_, ?_⟩ hc
+      · exact 𝟙 c.pt
+      · rintro (⟨⟩|⟨⟩) <;> simp
+      · exact 𝟙 c.pt
+      · rintro (⟨⟩|⟨⟩) <;> simp
+      · ext; simp
+      · ext; simp
+
+    refine Limits.IsColimit.equivOfNatIsoOfIso
+      ⟨⟨fun | .zero => 𝟙 _ | .one => 𝟙 _, ?_⟩,
+        ⟨fun | .zero => 𝟙 _ | .one => 𝟙 _, ?_⟩, ?_, ?_⟩ _ _
+        ⟨⟨?_, ?_⟩, ⟨?_, ?_⟩, ?_, ?_⟩ $
+        ShortComplex.exact_and_epi_g_iff_g_is_cokernel (s.map F) |>.1
+          (h s ⟨exact0, epi0⟩) |>.some
+    · rintro _ _ (⟨⟩ | ⟨⟩ | ⟨_⟩) <;> simp
+    · rintro _ _ (⟨⟩ | ⟨⟩ | ⟨_⟩) <;> simp
+    · ext (⟨⟩|⟨⟩) <;> simp
+    · ext (⟨⟩|⟨⟩) <;> simp
+    · exact 𝟙 _
+    · rintro (⟨⟩ | ⟨⟩) <;> simp
+    · exact 𝟙 _
+    · rintro (⟨⟩ | ⟨⟩) <;> simp
+    · ext; simp
+    · ext; simp
+
+  tfae_have 3 → 4
+  · intro h; refine ⟨?_⟩
+    apply (config := {allowSynthFailures := true}) preservesFiniteColimitsOfPreservesCokernels
+    exact fun {X Y} f => (h f).some
+
+  tfae_have 4 → 1
+  · rintro ⟨inst⟩ S hS
+    refine (S.map F).exact_and_epi_g_iff_g_is_cokernel |>.2 ⟨?_⟩
+    have := S.exact_and_epi_g_iff_g_is_cokernel.1 ⟨hS.exact, hS.epi_g⟩ |>.some
+    have := isColimitOfPreserves F this
+    refine Limits.IsColimit.equivOfNatIsoOfIso ?_ _ _ ?_ this
+    · refine NatIso.ofComponents (fun
+      | .zero => Iso.refl _
+      | .one => Iso.refl _) ?_
+      · rintro (_|_) (_|_) (_|_|_) <;> simp
+    · refine ⟨?_, ?_, ?_, ?_⟩
+      · refine ⟨𝟙 _, ?_⟩
+        rintro (_|_) <;> simp [← F.map_comp]
+      · refine ⟨𝟙 _, ?_⟩
+        rintro (_|_) <;> simp [← F.map_comp]
+      · ext; simp
+      · ext; simp
+
+  tfae_finish
+
+open ZeroObject in
+lemma preserves_shortComplex_exact_iff_preserves_finite_limit_colimit :
```

Comment:
Here, you should make a `preservesHomology_tfae` (or `exact_tfae`?) which would say that it is equivalent to preserve `ShortExact`, to preserve `Exact`, to `PreservesHomology` and the combination of the preservation of finite limits/colimits. You should use `preservesFiniteLimitsOfPreservesHomology`, `preservesFiniteLimitsOfPreservesHomology` (the other direction is an instance), and also `preservesHomologyOfMapExact`.

Reply:
Thanks for the suggestions and the code golfing!!